### PR TITLE
Add several updates to `vim._with` (tests, granular option contexts, `env` context)

### LINKED ISSUE #29253: `vim._with()` does not properly restore global-local options
### Problem

When using `options` context in `vim._with()`, it can result into unwanted side effects.

### Steps to reproduce

1. Create the 'script.lua' file with the following content:
    ```lua
    vim.go.commentstring, vim.bo.commentstring = '- %s', '-- %s'

    _G.log = {}
    table.insert(_G.log, { 'Before', go = vim.go.commentstring, bo = vim.bo.commentstring })
    vim._with({ options = { commentstring = '## %s' } }, function()
      table.insert(_G.log, { 'Inside', go = vim.go.commentstring, bo = vim.bo.commentstring })
    end)
    table.insert(_G.log, { 'After', go = vim.go.commentstring, bo = vim.bo.commentstring })
    ```
2. `nvim --clean -- script.lua`
3. `:source`
4. `:=log`.

The output now (notice difference in `go` values):

```
:=log
{ { "Before",
    bo = "-- %s",
    go = "- %s"
  }, { "Inside",
    bo = "## %s",
    go = "## %s"
  }, { "After",
    bo = "-- %s",
    go = "-- %s"
  } }
```

### Expected behavior

Both global and local option values are the same after executing `vim._with()` as they were before it.

### Neovim version (nvim -v)

NVIM v0.11.0-dev-200+g9afa1fd35

### Vim (not Nvim) behaves the same?

No (doesn't have this functionality)

### Operating system/version

EndeavourOS Linux x86_64 (6.9.3-arch1-1)

### Terminal name/version

st-0.9

### $TERM environment variable

st-256color

### Installation

appimage

### PR DESCRIPTION
**Pull request by echasnovski**
This PR evolved into three improvements:

- *Massively* cover existing functionality with tests. Some (many) tests are marked as pending because the particular context does not work. Those tests are meant as a reference for future fixing.
- Replace `options` context with `o`, `bo`, `wo`, `go` granular contexts. This aligns with existing Lua options terminology and is more flexible than single `options`.
- Add `env` context for environment variables. As this also aligns with `vim.env`.

More details in commits' messages.

------

Resolves #29253

Alternative to #29269 and #29268

### CODE REVIEW

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -1151,34 +1154,70 @@ end
...
 --- @field noautocmd? boolean
---- @field options? table<string, any>
+--- @field o? table<string, any>
 --- @field sandbox? boolean
 --- @field silent? boolean
 --- @field unsilent? boolean
 --- @field win? integer
+--- @field wo? table<string, any>
+
+local save_set_context = function(context)
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua:1164]: The name might need a bit of explanation, is this correct:
```suggestion
--- Gets option :set context.
local save_set_context = function(context)
```

Nit: "save" implies side-effects, so "get" would normally be preferrable, but `get_set_context` would be strange.

Comment (justinmk) [reply runtime/lua/vim/shared.lua:1164]: > How about `cache_and_set_context()`?

doesn't seem better than "save"

Comment (justinmk) [reply runtime/lua/vim/shared.lua:1164]: > How about `cache_and_set_context()`?

doesn't seem better than "save".  I see now that this is setting options so it does in fact have side effects, so ignore my previous comment. Just maybe a brief docstring would help.

File: runtime/lua/vim/shared.lua [hunk 2]
```diff
@@ -1151,34 +1154,70 @@ end
...
+    ['context.keeppatterns'] = { context.keeppatterns, 'boolean', true },
+    ['context.lockmarks'] = { context.lockmarks, 'boolean', true },
+    ['context.noautocmd'] = { context.noautocmd, 'boolean', true },
+    ['context.o'] = { context.o, 'table', true },
+    ['context.sandbox'] = { context.sandbox, 'boolean', true },
+    ['context.silent'] = { context.silent, 'boolean', true },
+    ['context.unsilent'] = { context.unsilent, 'boolean', true },
+    ['context.win'] = { context.win, 'number', true },
+    ['context.wo'] = { context.wo, 'table', true },
+  })
```
Comment (gpanders) [inline runtime/lua/vim/shared.lua:1220]: Out of curiosity, is there a reason to change this?

The single-line `vim.validate` calls are basically free. I don't think changing this is a readability win either way (there's a small argument to be made for eliminating the repeated `vim.validate`, but that's it). My main concern is about establishing patterns. Future contributors will copy-paste what they find elsewhere, and I think we should try to use the fast version of `vim.validate` wherever possible for that reason.

Comment (echasnovski) [reply runtime/lua/vim/shared.lua:1220]: My bad, it was more of an assumption that single call to `vim.validate()` is more performant than multiple. However, I did not account that those are the newly added style of vararg calls instead of creating a table every single time.

I've profiled two versions (single `vim.validate()` with table argument vs multiple with vararg) and indeed vararg version seems to be faster: around 5e-07 seconds for single `vim.validate()` and around 8e-08 seconds for vararg (time for all current 19 calls).

I'll change back to multiple vararg calls.

File: runtime/lua/vim/shared.lua (hunk 1 above)
Comment (echasnovski) [reply runtime/lua/vim/shared.lua]: I resolved this by compactifying implementation so that it can be reasonably fit inside `vim._with()` body. No helper function - no problem.

File: runtime/lua/vim/shared.lua [hunk 3]
```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
...
+
+    -- Use specific order from possibly most to least intrusive
+    for _, scope in ipairs(scope_order) do
+      for name, context_value in pairs(context[scope] or {}) do
+        local sc = scope == 'o' and scope_map[vim.api.nvim_get_option_info2(name, {}).scope]
+          or scope
+
+        -- Do not override already set cache and fall back to `vim.NIL` to
+        -- cache `nil` values (which needs restoring later)
+        cache[sc][name] = cache[sc][name] or vim[sc][name] or vim.NIL
```
Comment (dundargoc) [inline runtime/lua/vim/shared.lua:1243]: Do I understand it correctly that e.g. if both `o` and `bo` scopes are used with different values then `o` will be used since it has higher priority?

**Review by dundargoc — APPROVED** [review]
I haven't checked the test code but I assume they hold the standard echa quality. The option code is somewhat hard to follow, but I'm unsure if there's a good way to simplify it tbh. Maybe we can look into that later. All in all looks great. All the test cases are super valuable.

File: runtime/lua/vim/shared.lua [hunk 4]
```diff
@@ -1151,21 +1154,46 @@ end
...
   vim.validate('context', context, 'table')
   vim.validate('f', f, 'function')
 
+  vim.validate('context.bo', context.bo, 'table', true)
   vim.validate('context.buf', context.buf, 'number', true)
   vim.validate('context.emsg_silent', context.emsg_silent, 'boolean', true)
+  vim.validate('context.env', context.env, 'table', true)
+  vim.validate('context.go', context.go, 'table', true)
   vim.validate('context.hide', context.hide, 'boolean', true)
   vim.validate('context.horizontal', context.horizontal, 'boolean', true)
```
Comment (dundargoc) [inline runtime/lua/vim/shared.lua:1198]: This is an error from my part. This context does not exists in `vim._with_c` and should probably be removed for the time being.

File: runtime/lua/vim/shared.lua [hunk 5]
```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
...
-      previous_options[k] =
-        vim.api.nvim_get_option_value(k, { win = context.win, buf = context.buf })
-      vim.api.nvim_set_option_value(k, v, { win = context.win, buf = context.buf })
+  -- Decorate so that save-set-restore options is done in correct window-buffer
+  local callback = function()
+    -- Save and set context (NOTE: triggers `OptionSet` event)
+    local cache = { bo = {}, env = {}, go = {}, wo = {} } ---@type table<string, any>
+
+    -- Use specific order from possibly most to least intrusive
+    for _, scope in ipairs(scope_order) do
```
[the remark this answers is not in the document]
Comment (echasnovski) [reply runtime/lua/vim/shared.lua:1236]: I am not sure how to make it less confusing, as it is caching and setting inside for loops.

Here is an example of what it does if it would make it easier to follow. For `{ o = { winblend = 25 }, bo = { undolevels = 0 } }`, it does the following sequence:
- Process `o` context:
    - Get its "effective"/"cachable" context via `nvim_get_option_info2()`. The result is one of `bo`, `go`, or `wo`. It is `wo` for `winblend`.
    - Cache it via `cache.wo.winblend = cache.wo.winblend or vim.wo.winblend or vim.NIL`. The first `cache.wo.winblend` is there in order to write to cache only once (as the same option name can be supplied in three different `context` fields). The fallback to `vim.NIL` is there mostly for `env` context (as `vim.env` can return `nil`).
    - Cache its global value via `cache.go.winblend = cache.go.winblend or vim.go.winblend`. This is needed because later setting option might affect global values (like for `o` or `wo` contexts).
    - Set it with `vim.o.winblend = 25`.
- Process `bo` context:
    - `cache.bo.undolevels = cache.bo.undolevels or vim.bo.undolevels or vim.NIL`.
    - `cache.go.undolevels = cache.go.undolevels or vim.go.undolevels`.
    - `vim.bo.undolevels = 0`.
- Execute callback `f`.
- Restore cache:
    - `vim.bo.undolevels = cache.bo.undolevels`
    - `vim.wo.winblend = cache.wo.winblend`
    - `vim.go.undolevels = cache.go.undolevels` and `vim.go.winblend = cache.go.winblend`.

File: runtime/lua/vim/shared.lua (hunk 3 above)
Comment (echasnovski) [reply runtime/lua/vim/shared.lua:1243]: On the contrary, `bo` will be used as it will be set *after* setting the `o` context.

Using `cache[sc][name] = cache[sc][name] or vim[sc][name]` instead of direct `cache[sc][name] = vim[sc][name]` guards exactly from this kind of case. Without it, cache would have been set to user's value from the `o` context and later (erroneously) restored to it.

File: runtime/lua/vim/shared.lua (hunk 4 above)
Comment (echasnovski) [reply runtime/lua/vim/shared.lua]: I removed it from code, but left pending tests just in case.

File: runtime/lua/vim/shared.lua (hunk 3 above)
Comment (echasnovski) [reply runtime/lua/vim/shared.lua:1241]: But good question, as I realized that there was no test case specifically if same option is supplied in several contexts.

File: runtime/lua/vim/shared.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (dundargoc) [reply runtime/lua/vim/shared.lua:1234]: Ah ok, that clears things up a bit. Hmm, yeah I can't come up with a simpler approach for the time being. Let's go with this 👍

File: runtime/lua/vim/shared.lua (hunk 1 above)
Comment (echasnovski) [reply runtime/lua/vim/shared.lua]: I've recently had a realization that if there is a misplaced option scope (like `{ bo = { winblend = 10 } }`), then the approach of "cache and set" inside same double loop would not properly restore already set context values (due to error during `vim.bo.winblend = 10`). So I indeed had to separate `get_context_state()` into a separate function (with brief description of what it does).

File: runtime/lua/vim/shared.lua [hunk 6]
```diff
@@ -1141,44 +1141,105 @@ end
...
+--- @field bo? table<string, any>
+--- @field env? table<string, any>
+--- @field go? table<string, any>
+--- @field wo? table<string, any>
+
+local scope_map = { buf = 'bo', global = 'go', win = 'wo' }
+local scope_order = { 'o', 'wo', 'bo', 'go', 'env' }
+local state_restore_order = { 'bo', 'wo', 'go', 'env' }
+
+--- Get data about current state, enough to properly undo later context setting
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua:1174]:
```suggestion
--- Gets data about current state, enough to properly restore specified options/env/etc.
```

File: test/functional/lua/with_spec.lua [hunk 7]
```diff
@@ -7,286 +7,1620 @@ local api = n.api
...
     ]]
-    t.eq(mark, { 1, 0 })
+    local get_error = function(string_args)
+      return exec_lua('return get_error(' .. string_args .. ')')
+    end
+
+    matches('context.*table', get_error("'a', function() end"))
+    matches('f.*function', get_error('{}, 1'))
+
+    local validate_context = function(bad_context, expected_type)
```
Comment (justinmk) [inline test/functional/lua/with_spec.lua:1596]: we are fairly consistent about using `assert_` prefix for tests
```suggestion
    local assert_context = function(bad_context, expected_type)
```

**Review by justinmk — APPROVED** [review]

**Comment by dundargoc** [comment]
This is super valuable. Thanks for championing this.

Follow-up work (not necessarily by you, I'm talking in general): ensure all the tests marked as `pending` passes. It looks like it involved primarily ensuring that `vim._with` can be nested which will need adjustments in the C code.

**Comment by echasnovski** [comment]
> Follow-up work (not necessarily by you, I'm talking in general): ensure all the tests marked as `pending` passes. It looks like it involved primarily ensuring that `vim._with` can be nested which will need adjustments in the C code.

Not only those, I am afraid. Quite a few context fields coming from command modifiers don't seem to work:
- [`emsg_silent`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L306)
- [`hide`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L512)
- [`horizontal`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L565) (although it is not even implemented in C part)
- [`keepalt`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L591)
- [`keepjumps`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L625)
- [`keepmarks`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L654)
- [`keeppatterns`](https://github.com/neovim/neovim/blob/76dd07e572935742f6dfa0d3364e262ebc047dcc/test/functional/lua/with_spec.lua#L687)

Soo... basically most part of the flags :)
